remove RTTI check for TensorImpl shadow copy

ljk53 opened the pull request:
Summary:
We introduced RTTI in recent change: https://github.com/pytorch/pytorch/pull/21613

For internal mobile build we don't enable '-frtti' yet. This diff is trying to replace
RTTI with alternative approach.

According to dzhulgakov we could compare two tensors' type_id directly in most cases -
which is more strict than comparing TensorImpl subclass type as TensorImpl -> type_id
mapping is 1-to-n but it's more proper for this use case.

The only two cases where we can relax direct type comparison (for legacy reason) are:
1. CPUTensor <-> CUDATensor;
2. SparseCPUTensor <-> SparseCUDATensor;

Differential Revision: D16212472

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2119,7 +2119,7 @@
   variants: method
 
-- func: _has_same_tensorimpl_type(Tensor self, Tensor other) -> bool
+- func: _has_compatible_type_for_copy(Tensor self, Tensor other) -> bool
```
1) I know you already prefixed this with "_", but this name makes it sound like it's (data) copy related, when it's really shallow copy related.  I'd just name this to something scarier, like `_has_compatible_shallow_copy_type` or something (just something mentioning either shallow copy or tensorimpl is sufficient).
2) Can you rename `other` to `from` or `to`?  I know it doesn't matter in the implementation for now, but I know if I see the name "same type" (as before) it doesn't matter what order I pass the parameters, but if I see the "copy" I immediately look what is "from" or "to" and if it's not obvious from the names (like this), I have to look up the implementation.

gchanan commented on c10/core/TensorImpl.h:
```diff
@@ -129,6 +129,20 @@ struct C10_API PlacementDeleteContext {
+      ((self == CPUTensorId() || self == CUDATensorId()) &&
+      (other == CPUTensorId() || other == CUDATensorId())) ||
+      ((self == SparseCPUTensorId() || self == SparseCUDATensorId()) &&
+      (other == SparseCPUTensorId() || other == SparseCUDATensorId()));
```
I guess this would technically break complex cpu/cuda tensors, but that's probably okay.

ljk53 replied:
@dzhulgakov @yf225 can comment? I don't have enough context to determine whether we should add complex cpu/cuda pair here... (according to dzhulgakov the current two pairs are for legacy reason so adding them here as special rule instead of some more generic condition based on is_sparse/is_cpu/is_cuda to cover more similar cases).

gchanan replied:
I believe the full story is that in prior releases there was an extra layer of indirection so doing:
`tensor.data = other` was legal for any tensor types for both `tensor` and `other`.

There's no longer that extra level of indirection, so we changed the support to only allow you to do this if it works to do a shallow copy (i.e. same TensorImpl type).  We haven't done a release with this behavior.

Furthermore, complex types is an extension (i.e. you have to install it separately), so in theory someone could be doing the above with `tensor` as CPU complex and `other` as CUDA complex, but that seems pretty unlikely given all the moving pieces here.

ljk53 replied:
@gchanan - thanks for the context. yeah - I wasn't sure how far we'd like to go to limit cross-type shallow-copy. I'm using strict type comparison + 2 white-list cases here which is more strict than already-tightened-same-TensorImpl-type constraint. It makes more sense if the 2-white-list cases are relatively 'fixed' (they are allowed only for some legacy reason). If we need keep extending it as we introduce new type (as @bddppq commented below for HIPTensorId case) I feel it's a little bit suboptimal.
@dzhulgakov your thoughts? do you feel HIPTensorId is the only case we should add or there will be more legit cases we need add here?

gchanan commented on c10/core/TensorImpl.h:
```diff
@@ -129,6 +129,20 @@ struct C10_API PlacementDeleteContext {
+// type_id. The only two special cases (for legacy reason) are:
+// CPUTensorId is compatible with CUDATensorId and SparseCPUTensorId is
+// compatible with SparseCUDATensorId.
+inline bool has_compatible_type_for_copy(
```
can we move this somewhere else?  Either right next to shallow_copy_from (okay, it doesn't really belong there because it's not TensorImpl related, but you always call them together), or to where we usually do checking/translation, which I think is currently in `Backend.h` (those should probably be TensorTypeId based, but oh well...).

ljk53 replied:
moved into TensorImpl as inline method

gchanan commented on c10/core/TensorImpl.h:
```diff
@@ -907,6 +907,21 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
+  inline bool has_compatible_shallow_copy_type(const TensorTypeId& from) {
+    const TensorTypeId& self = type_id();
+    return (self == from) ||
+        ((self == CPUTensorId() || self == CUDATensorId()) &&
```
I'm not sure I fully understand how HIP fits in to this, but I _think_ a HIP tenosr will have `HIPTensorId` even though its device is cuda and we want equivalent hip and cuda support.  I think @bddppq would know for sure.

bddppq replied:
Here should add `self == HIPTensorId()` check as well (and similar for the other checks below)

gchanan reviewed (approved):
approved assuming we figure out the correct thing to do for HIP.

ljk53 commented:
@pytorchbot retest this please

ljk53 committed:
remove RTTI check for TensorImpl shadow copy

Summary:
We introduced RTTI in recent change: https://github.com/pytorch/pytorch/pull/21613

For internal mobile build we don't enable '-frtti' yet. This diff is trying to replace
RTTI with alternative approach.

According to dzhulgakov we could compare two tensors' type_id directly in most cases -
which is more strict than comparing TensorImpl subclass type as TensorImpl -> type_id
mapping is 1-to-n but it's more proper for this use case.

The only two cases where we can relax direct type comparison (for legacy reason) are:
1. CPUTensor <-> CUDATensor;
2. SparseCPUTensor <-> SparseCUDATensor;

Differential Revision: D16212472

fbshipit-source-id: 5946ca605e86820329762f84761db9142fd06a29

gchanan commented:
@ljk53 are you not using gh tools?
